Environment workflows creation and refactor

craciunoiuc opened the pull request:
This PR does multiple things:
 * Bump gcc dockerfile to `12.2.0`;
 * Reformat existing actions to work offline with `act`;
 * Create `base` dockerfile to package `kraft`;
 * Add workflows to build `base`, and `myself`;
 * Bump qemu to `7.1.0`;
 * Create makefile to build environments;

NOTE: Building `base.Dockerfile` will fail as the FROM statements need to be pinned to the correct images. One method would be to inject them as build arguments, but not sure yet.

Depends: #210

Closes: #96
Closes: #97
Closes: #98
Closes: #2
Closes: #3
Closes: #143

craciunoiuc committed:
Makefile: Add option to also clean packages

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
buildenvs: Bump gcc to version '12.2.0'

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
buildenvs: Bump qemu to version '7.1.0'

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
buildenvs: Update myself environment

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
.dockerignore: Ignore files when building docker containers

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
buildenvs: Add 'base' container to run kraftkit

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
.github/workflows: Add workflow to build 'myself'

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
.github/workflows: Add workflow to build the 'base' container

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
buildenvs/Makefile: Add rules for environments

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
.github/workflows: Add workflow to check kraftkit building

GitHub-Fixes: #143
Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc commented:
Should be good to go after #181 is merged

nderjung commented on buildenvs/myself/kraftkit_pipeline_config.yaml:
```diff
@@ -0,0 +1,21 @@
+no_prompt: true
+no_parallel: true
+no_emojis: true
+git_protocol: https
+runtime_dir: /var/kraftkit
+default_plat: ""
+default_arch: ""
+events_pidfile: /var/kraftkit/events.pid
+paths:
+  plugins: /root/.local/share/kraftkit/plugins
+  config: /root/.config/kraftkit
+  manifests: /root/.local/share/kraftkit/manifests
+  sources: /root/.local/share/kraftkit/sources
+log:
+  level: warn
+  timestamps: false
+  type: none
+unikraft:
+  manifests:
+      - https://github.com/unikraft/app-*
+      - https://manifests.kraftkit.sh/index.yaml
```
This file can be removed imho

nderjung commented on buildenvs/myself.Dockerfile:
```diff
@@ -40,13 +14,27 @@ ARG GO_VERSION=${GO_VERSION}
+    make kraft; \
+    kraft -h;
+
+# Pull manifests - uses workaround
+RUN set -xe; \
+    true | (setsid kraft pkg update) 2>&1 | cat; \
+    sed -i 's/none/fancy/g' "${HOME}/.config/kraftkit/config.yaml";
```
Also can be removed

craciunoiuc replied:
can be done with commands now

also the first one, so removing

nderjung commented on buildenvs/gcc.Dockerfile:
```diff
@@ -130,10 +101,11 @@ RUN set -ex; \
     GCC_CONFIGURE_ARGS="$GCC_CONFIGURE_ARGS --target=${GCC_PREFIX}"; \
     ../gcc-${GCC_VERSION}/configure ${GCC_CONFIGURE_ARGS}; \
-    make all-gcc; \
+    make -j$(($(nproc) - 1)) all-gcc; \
```
This is a bit dangerous to execute since it will consume all the resources of a machine. Perhaps use an ARG instead, e.g. `MAKE_NPROC` so this can be turned off, it's default value should be `1`, so:
```diff
+ ARG MAKE_NPROC=1
[...]
- make all-gcc; \
+ make -j${MAKE_NPROC} all-gcc; \
```

nderjung commented on buildenvs/base.Dockerfile:
```diff
@@ -0,0 +1,151 @@
+COPY --from=gcc-x86_64 /libexec/gcc/x86_64-linux-gnu/12.2.0/collect2 /libexec/gcc/x86_64-linux-gnu/12.2.0/collect2
+# COPY the others if needed
+
+# COPY --from=gcc-arm /bin/                                            /bin
+# COPY --from=gcc-arm /lib/gcc/                                        /lib/gcc
+# COPY --from=gcc-arm /include/                                        /include
+# COPY --from=gcc-arm /lib/gcc/                                        /lib/gcc
+# COPY --from=gcc-arm /arm-linux-gnueabihf                             /arm-linux-gnueabihf
+# COPY --from=gcc-arm /libexec/gcc/arm-linux-gnueabihf/12.2.0/cc1      /libexec/gcc/arm-linux-gnueabihf/12.2.0/cc1
+# COPY --from=gcc-arm /libexec/gcc/arm-linux-gnueabihf/12.2.0/collect2 /libexec/gcc/arm-linux-gnueabihf/12.2.0/collect2
+# COPY the others if needed
+
+# COPY --from=gcc-arm64 /bin/                                          /bin
+# COPY --from=gcc-arm64 /lib/gcc/                                      /lib/gcc
+# COPY --from=gcc-arm64 /include/                                      /include
+# COPY --from=gcc-arm64 /lib/gcc/                                      /lib/gcc
+# COPY --from=gcc-arm64 /aarch64-linux-gnu/                            /aarch64-linux-gnu
+# COPY --from=gcc-arm64 /libexec/gcc/aarch64-linux-gnu/12.2.0/cc1      /libexec/gcc/aarch64-linux-gnu/12.2.0/cc1
+# COPY --from=gcc-arm64 /libexec/gcc/aarch64-linux-gnu/12.2.0/collect2 /libexec/gcc/aarch64-linux-gnu/12.2.0/collect2
+# COPY the others if needed
```
Eventually these will be added, why aren't they included atm? I forgot

craciunoiuc replied:
well, first of all they are not available

Also, we are currently not compiling `gcc` for those arches

Also not sure how well kraftkit works with `arm`

I suggest we leave them like this for now

nderjung commented on buildenvs/Makefile:
```diff
@@ -0,0 +1,88 @@
+		$(DOCKER_BUILD_EXTRA) $(WORKDIR)
+
+.PHONY: myself
+myself: GO_VERSION         ?= 1.18
+myself: ENVIRONMENT        ?= myself
+myself: IMAGE              ?= $(REG)/$(ORG)/$(APP)/$(ENVIRONMENT):$(IMAGE_TAG)
+myself:
```
Reduce indentation

nderjung commented on buildenvs/Makefile:
```diff
@@ -0,0 +1,88 @@
+		$(DOCKER_BUILD_EXTRA) $(WORKDIR)
+
+.PHONY: base
+base: GO_VERSION         ?= 1.18
+base: UK_ARCH            ?= x86_64
+base: GCC_VERSION        ?= 12.2.0
+base: QEMU_VERSION       ?= 7.1.0
+base: GCC_PREFIX         ?= x86_64-linux-gnu
+base: ENVIRONMENT        ?= base
+base: IMAGE              ?= $(REG)/$(ORG)/$(APP)/$(ENVIRONMENT):$(IMAGE_TAG)
```
Reduce indentation

nderjung commented on Makefile:
```diff
@@ -156,6 +139,10 @@ cicheck:
 clean:
 	$(GO) clean -cache -i -r
 
+.PHONY: pkgclean
+pkgclean:
+	$(GO) clean -cache -i -r -modcache
+
```
I'd add `-modcache` `clean` target as opposed to new target

nderjung commented on .github/workflows/myself.yaml:
```diff
@@ -0,0 +1,38 @@
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+
+      - name: Login to DockerHub
+        uses: docker/login-action@v2
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
```
To be removed, we should use our own registry :)

craciunoiuc replied:
I'll leave this in for now, till our own registry is ready

nderjung commented on .github/workflows/base.yaml:
```diff
@@ -0,0 +1,144 @@
+jobs:
+  gcc-x86_64:
+    runs-on: ubuntu-latest
+    env:
+      GCC_VERSION: 12.2.0
+      GCC_ARCH: x86_64
+      GCC_SUFFIX:
+
```
Perhaps we can use a matrix, e.g.:
```yaml
jobs:
  archs:
    strategy:
      matrix:
        include:
          - GCC_VERSION: 12.2.0
            GCC_ARCH: x86_64
          - GCC_VERSION: 12.2.0
            GCC_ARCH: arm
          - GCC_VERSION: 12.2.0
            GCC_ARCH: arm64
```

craciunoiuc replied:
yes, we actually should, but right now all other components are disabled.

With only two elements or less I don't see a reason to use a matrix yet

Once we'll support all 3 I think we should add it

nderjung replied:
Can you set it up to use the matrix and then delete the commented-out `gcc-arm`/`gcc-arm64` and just comment out the matrix entries?

craciunoiuc replied:
pray it works :wink:

nderjung commented on .github/workflows/base.yaml:
```diff
@@ -0,0 +1,144 @@
+        uses: docker/build-push-action@v3
+        with:
+          file: ./buildenvs/myself.Dockerfile
+          tags: kraftkit/myself:latest
```
```diff
- tags: kraftkit/myself:latest
+ tags: kraftkit.sh/myself:latest
```

nderjung commented on .github/workflows/gobuild.yaml:
```diff
@@ -0,0 +1,21 @@
+jobs:
+  unit:
+    runs-on: ubuntu-latest
+    container: unikraft/kraftkit/myself:latest
```
```diff
- container: unikraft/kraftkit/myself:latest
+ container: kraftkit.sh/myself:latest
```

nderjung commented on .github/workflows/base.yaml:
```diff
@@ -0,0 +1,103 @@
+        uses: docker/build-push-action@v3
+        with:
+          push: true
+          tags: kraftkit/base:latest
```
```diff
- tags: kraftkit/base:latest
+ tags: kraftkit.sh/base:latest
```

nderjung commented on .github/workflows/myself.yaml:
```diff
@@ -0,0 +1,39 @@
+          push: true
+          file: ./buildenvs/myself.Dockerfile
+          platforms: linux/amd64
+          tags: kraftkit/myself:latest
```
```diff
- tags: kraftkit/myself:latest
+ tags: kraftkit.sh/myself:latest
```

nderjung commented on buildenvs/Makefile:
```diff
@@ -0,0 +1,92 @@
+BUILDENVSDIR := $(patsubst %/,%,$(dir $(_MKFILE_PATH)))
+WORKDIR      ?= $(BUILDENVSDIR)/..
+
+REG          ?= ghcr.io
+ORG          ?= unikraft
```
```diff
- REG          ?= ghcr.io
+ REG          ?= kraftkit.sh
- ORG          ?= unikraft
```

nderjung commented on buildenvs/Makefile:
```diff
@@ -0,0 +1,92 @@
+qemu: QEMU_VERSION ?= 7.1.0
+qemu: MAKE_NPROC   ?= $(shell nproc)
+qemu: ENVIRONMENT  ?= qemu
+qemu: IMAGE        ?= $(REG)/$(ORG)/$(APP)/$(ENVIRONMENT):$(QEMU_VERSION)
```
```diff
- qemu: IMAGE        ?= $(REG)/$(ORG)/$(APP)/$(ENVIRONMENT):$(QEMU_VERSION)
+ qemu: IMAGE        ?= $(REG)/$(APP)/$(ENVIRONMENT):$(QEMU_VERSION)
```

nderjung commented on buildenvs/Makefile:
```diff
@@ -0,0 +1,92 @@
+gcc: BINUTILS_VERSION ?= 2.39
+gcc: GLIB_VERSION     ?= 2.31
+gcc: ENVIRONMENT      ?= gcc
+gcc: IMAGE            ?= $(REG)/$(ORG)/$(APP)/$(ENVIRONMENT):$(GCC_VERSION)-$(UK_ARCH)
```
```diff
- gcc: IMAGE            ?= $(REG)/$(ORG)/$(APP)/$(ENVIRONMENT):$(GCC_VERSION)-$(UK_ARCH)
+ gcc: IMAGE            ?= $(REG)/$(APP)/$(ENVIRONMENT):$(GCC_VERSION)-$(UK_ARCH)
```

nderjung commented on buildenvs/Makefile:
```diff
@@ -0,0 +1,92 @@
+.PHONY: myself
+myself: GO_VERSION  ?= 1.18
+myself: ENVIRONMENT ?= myself
+myself: IMAGE       ?= $(REG)/$(ORG)/$(APP)/$(ENVIRONMENT):$(IMAGE_TAG)
```
```diff
- myself: IMAGE       ?= $(REG)/$(ORG)/$(APP)/$(ENVIRONMENT):$(IMAGE_TAG)
+ myself: IMAGE       ?= $(REG)/$(APP)/$(ENVIRONMENT):$(IMAGE_TAG)
```

nderjung commented on buildenvs/Makefile:
```diff
@@ -0,0 +1,92 @@
+base: GCC_PREFIX   ?= x86_64-linux-gnu
+base: QEMU_VERSION ?= 7.1.0
+base: ENVIRONMENT  ?= base
+base: IMAGE        ?= $(REG)/$(ORG)/$(APP)/$(ENVIRONMENT):$(IMAGE_TAG)
```
```diff
- base: IMAGE        ?= $(REG)/$(ORG)/$(APP)/$(ENVIRONMENT):$(IMAGE_TAG)
+ base: IMAGE        ?= $(REG)/$(APP)/$(ENVIRONMENT):$(IMAGE_TAG)
```

nderjung commented on buildenvs/base.Dockerfile:
```diff
@@ -0,0 +1,149 @@
+
+ARG GCC_VERSION=12.2.0
+ARG GCC_SUFFIX=
+
```
```diff
+ ARG REG=kraftkit.sh
```

nderjung commented on buildenvs/base.Dockerfile:
```diff
@@ -0,0 +1,149 @@
+
+ARG DEBIAN_VERSION=bullseye-20221114
+
+FROM ${REG}/unikraft/kraftkit/gcc:${GCC_VERSION}-x86_64${GCC_SUFFIX} AS gcc-x86_64
```
```diff
- FROM ${REG}/unikraft/kraftkit/gcc:${GCC_VERSION}-x86_64${GCC_SUFFIX} AS gcc-x86_64
+ FROM ${REG}/gcc:${GCC_VERSION}-x86_64${GCC_SUFFIX} AS gcc-x86_64
```

nderjung commented on buildenvs/base.Dockerfile:
```diff
@@ -0,0 +1,149 @@
+# FROM ${REG}/unikraft/kraftkit/gcc:${GCC_VERSION}-arm64${GCC_SUFFIX}  AS gcc-arm64
+
+FROM ${REG}/unikraft/kraftkit/qemu:${QEMU_VERSION}                   AS qemu
+FROM ${REG}/unikraft/kraftkit/myself:${KRAFTKIT_VERSION}             AS kraft
```
```diff
- FROM ${REG}/unikraft/kraftkit/gcc:${GCC_VERSION}-x86_64${GCC_SUFFIX} AS gcc-x86_64
- # FROM ${REG}/unikraft/kraftkit/gcc:${GCC_VERSION}-arm${GCC_SUFFIX}    AS gcc-arm
- # FROM ${REG}/unikraft/kraftkit/gcc:${GCC_VERSION}-arm64${GCC_SUFFIX}  AS gcc-arm64
+ FROM ${REG}/gcc:${GCC_VERSION}-x86_64${GCC_SUFFIX}  AS gcc-x86_64
+ # FROM ${REG}/gcc:${GCC_VERSION}-arm${GCC_SUFFIX}   AS gcc-arm
+ # FROM ${REG}/gcc:${GCC_VERSION}-arm64${GCC_SUFFIX} AS gcc-arm64

- FROM ${REG}/unikraft/kraftkit/qemu:${QEMU_VERSION}                   AS qemu
- FROM ${REG}/unikraft/kraftkit/myself:${KRAFTKIT_VERSION}             AS kraft
+ FROM ${REG}/qemu:${QEMU_VERSION}                    AS qemu
+ FROM ${REG}/myself:${KRAFTKIT_VERSION}              AS kraft
```

nderjung commented on buildenvs/Makefile:
```diff
@@ -0,0 +1,89 @@
+WORKDIR      ?= $(BUILDENVSDIR)/..
+
+REG          ?= kraftkit.sh
+APP          ?= kraftkit
```
```diff
- APP          ?= kraftkit
```

nderjung commented on buildenvs/Makefile:
```diff
@@ -0,0 +1,89 @@
+qemu: QEMU_VERSION ?= 7.1.0
+qemu: MAKE_NPROC   ?= $(shell nproc)
+qemu: ENVIRONMENT  ?= qemu
+qemu: IMAGE        ?= $(REG)/$(APP)/$(ENVIRONMENT):$(QEMU_VERSION)
```
```diff
- qemu: IMAGE        ?= $(REG)/$(APP)/$(ENVIRONMENT):$(QEMU_VERSION)
+ qemu: IMAGE        ?= $(REG)/$(ENVIRONMENT):$(QEMU_VERSION)
```

nderjung commented on buildenvs/Makefile:
```diff
@@ -0,0 +1,89 @@
+gcc: BINUTILS_VERSION ?= 2.39
+gcc: GLIB_VERSION     ?= 2.31
+gcc: ENVIRONMENT      ?= gcc
+gcc: IMAGE            ?= $(REG)/$(APP)/$(ENVIRONMENT):$(GCC_VERSION)-$(UK_ARCH)
```
```diff
- gcc: IMAGE            ?= $(REG)/$(APP)/$(ENVIRONMENT):$(GCC_VERSION)-$(UK_ARCH)
+ gcc: IMAGE            ?= $(REG)/$(ENVIRONMENT):$(GCC_VERSION)-$(UK_ARCH)
```

nderjung commented on buildenvs/Makefile:
```diff
@@ -0,0 +1,89 @@
+.PHONY: myself
+myself: GO_VERSION  ?= 1.20
+myself: ENVIRONMENT ?= myself
+myself: IMAGE       ?= $(REG)/$(APP)/$(ENVIRONMENT):$(IMAGE_TAG)
```
```diff
- myself: IMAGE       ?= $(REG)/$(APP)/$(ENVIRONMENT):$(IMAGE_TAG)
+ myself: IMAGE       ?= $(REG)/$(ENVIRONMENT):$(IMAGE_TAG)
```

nderjung commented on buildenvs/Makefile:
```diff
@@ -0,0 +1,89 @@
+base: GCC_PREFIX   ?= x86_64-linux-gnu
+base: QEMU_VERSION ?= 7.1.0
+base: ENVIRONMENT  ?= base
+base: IMAGE        ?= $(REG)/$(APP)/$(ENVIRONMENT):$(IMAGE_TAG)
```
```diff
- base: IMAGE        ?= $(REG)/$(APP)/$(ENVIRONMENT):$(IMAGE_TAG)
+ base: IMAGE        ?= $(REG)/$(ENVIRONMENT):$(IMAGE_TAG)
```

nderjung commented on buildenvs/Makefile:
```diff
@@ -0,0 +1,89 @@
+		--build-arg GCC_VERSION=$(GCC_VERSION) \
+		--build-arg GCC_PREFIX=$(GCC_PREFIX) \
+		--build-arg QEMU_VERSION=$(QEMU_VERSION) \
+		--build-arg APP=$(APP) \
```
```diff
- 		--build-arg APP=$(APP) \
```

nderjung commented on buildenvs/base.Dockerfile:
```diff
@@ -0,0 +1,149 @@
+
+# Unikraft workdir and environment
+WORKDIR /usr/src/unikraft/apps/app
+ENV UK_WORKDIR /usr/src/unikraft
+ENV UK_CACHEDIR /usr/src/unikraft/.kraftcache
+ENV KRAFTRC /usr/src/unikraft/.kraftrc
```
Legacy pykraft environmental variables
```diff
- ENV UK_WORKDIR /usr/src/unikraft
- ENV UK_CACHEDIR /usr/src/unikraft/.kraftcache
- ENV KRAFTRC /usr/src/unikraft/.kraftrc
```

nderjung commented on Makefile:
```diff
diff --git a/Makefile b/Makefile
index af1b89b..5d32780 100644
--- a/Makefile
+++ b/Makefile
@@ -11,7 +11,7 @@ INSTALLDIR  ?= /usr/local/bin/
 VENDORDIR   ?= $(WORKDIR)/vendor

 # Arguments
-REGISTRY    ?= ghcr.io
+REGISTRY    ?= kraftkit.sh
 ORG         ?= unikraft
 REPO        ?= kraftkit
 BIN         ?= kraft
 ```

nderjung replied:
```diff
diff --git a/Makefile b/Makefile
index af1b89b..5d32780 100644
--- a/Makefile
+++ b/Makefile
@@ -42,7 +42,7 @@ DOCKER_RUN  ?= $(DOCKER) run --rm $(1) \
                -e DOCKER= \
                -w /go/src/$(GOMOD) \
                -v $(WORKDIR):/go/src/$(GOMOD) \
-               $(REGISTRY)/$(ORG)/$(REPO)/$(2):$(IMAGE_TAG) \
+               $(REGISTRY)/$(2):$(IMAGE_TAG) \
                  $(3)
 GO          ?= go
 GOFUMPT     ?= gofumpt
 ```

nderjung replied:
```diff
diff --git a/Makefile b/Makefile
index af1b89b..5d32780 100644
--- a/Makefile
+++ b/Makefile
@@ -93,22 +93,10 @@ $(addprefix $(.PROXY), $(BIN)):
                -o $(DISTDIR)/$@ \
                $(WORKDIR)/cmd/$@

-# Create an environment where we can build
-.PHONY: container
-container: GO_VERSION         ?= 1.20
-container: DOCKER_BUILD_EXTRA ?=
-container: ENVIRONMENT        ?= myself
-container: IMAGE              ?= $(REGISTRY)/$(ORG)/$(REPO)/$(ENVIRONMENT):$(IMAGE_TAG)
-container: TARGET             ?= base
-container:
-       $(DOCKER) build \
-               --build-arg ORG=$(ORG) \
-               --build-arg REPO=$(REPO) \
-               --build-arg GO_VERSION=$(GO_VERSION) \
-               --tag $(IMAGE) \
-               --target $(TARGET) \
-               --file $(WORKDIR)/buildenvs/$(ENVIRONMENT).Dockerfile \
-               $(DOCKER_BUILD_EXTRA) $(WORKDIR)
+
+# Proxy all "build environment" (buildenvs) targets
+buildenv-%:
+       $(MAKE) -C $(WORKDIR)/buildenvs $*

 # Run an environment where we can build
 .PHONY: devenv
 ```

nderjung replied:
I'd include a separate commit explaining the simplification by proxying the targets.

nderjung commented on buildenvs/gcc.Dockerfile:
For some reason the official proxy is SUPER slow, let's switch to:
```diff
diff --git a/buildenvs/gcc.Dockerfile b/buildenvs/gcc.Dockerfile
index 9b3c0bb..80fd327 100644
--- a/buildenvs/gcc.Dockerfile
+++ b/buildenvs/gcc.Dockerfile
@@ -36,7 +36,7 @@ RUN set -ex; \
     apt-get clean; \
     mkdir -p ${PREFIX}/src; \
     cd ${PREFIX}/src; \
-    curl -O https://ftp.gnu.org/gnu/binutils/binutils-${BINUTILS_VERSION}.tar.gz; \
+    curl -O https://ftp.fu-berlin.de/unix/gnu/binutils//binutils-${BINUTILS_VERSION}.tar.gz; \
     tar zxf binutils-${BINUTILS_VERSION}.tar.gz; \
     rm binutils-${BINUTILS_VERSION}.tar.gz; \
     chown -R root:root binutils-${BINUTILS_VERSION}; \
@@ -66,7 +66,7 @@ RUN set -ex; \
     make; \
     make install; \
     cd ${PREFIX}/src; \
-    curl -O https://ftp.gnu.org/gnu/gcc/gcc-${GCC_VERSION}/gcc-${GCC_VERSION}.tar.gz; \
+    curl -O https://ftp.fu-berlin.de/unix/languages/gcc/releases/gcc-${GCC_VERSION}/gcc-${GCC_VERSION}.tar.gz; \
     tar zxf gcc-${GCC_VERSION}.tar.gz; \
     rm gcc-${GCC_VERSION}.tar.gz; \
     chown -R root:root gcc-${GCC_VERSION}; \
```

nderjung commented on buildenvs/base.Dockerfile:
```diff
@@ -0,0 +1,146 @@
+FROM ${REG}/qemu:${QEMU_VERSION}                   AS qemu
+FROM ${REG}/myself:${KRAFTKIT_VERSION}             AS kraft
+
+LABEL MAINTAINER="Alexander Jung <alexander.jung@unikraft.io>"
+LABEL MAINTAINER="Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>"
```
`LABEL MAINTAINER` is [deprecated](https://docs.docker.com/engine/reference/builder/#maintainer-deprecated) and can be removed.

nderjung commented on buildenvs/myself.Dockerfile:
```diff
@@ -14,14 +14,27 @@ ARG GO_VERSION=${GO_VERSION}
+COPY . .
+
 ENV GOROOT=/usr/local/go
 ENV PATH=$PATH:/go/src/kraftkit.sh/dist
```
```diff
diff --git a/buildenvs/myself.Dockerfile b/buildenvs/myself.Dockerfile
index 789ad53..308f00b 100644
--- a/buildenvs/myself.Dockerfile
+++ b/buildenvs/myself.Dockerfile
@@ -28,6 +28,7 @@ COPY . .

 ENV GOROOT=/usr/local/go
 ENV PATH=$PATH:/go/src/kraftkit.sh/dist
+ENV DOCKER=

 # Build the binary
 RUN set -xe; \
 ```

nderjung commented on .dockerignore:
```diff
@@ -0,0 +1,2 @@
+dist/
+.git/
```
DO include the `.git` folder as it is necessary in the top-level Makefile for generating version tags:
```diff
diff --git a/.dockerignore b/.dockerignore
index 43fe8fa..30b9507 100644
--- a/.dockerignore
+++ b/.dockerignore
@@ -1,2 +1,2 @@
 dist/
-.git/
\ No newline at end of file
+
```
Also include a newline to terminate the file.

nderjung commented on .github/workflows/base.yaml:
```diff
@@ -0,0 +1,103 @@
+        uses: docker/build-push-action@v3
+        with:
+          file: ./buildenvs/gcc.Dockerfile
+          tags: kraftkit.sh/gcc:${{ matrix.GCC_VERSION }}-${{ matrix.GCC_ARCH }}${{ env.GCC_SUFFIX }}
```
```diff
diff --git a/.github/workflows/base.yaml b/.github/workflows/base.yaml
index ea6b450..94c6c1b 100644
--- a/.github/workflows/base.yaml
+++ b/.github/workflows/base.yaml
@@ -33,7 +33,7 @@ jobs:
         uses: docker/build-push-action@v3
         with:
           file: ./buildenvs/gcc.Dockerfile
-          tags: kraftkit.sh/gcc:${{ matrix.GCC_VERSION }}-${{ matrix.GCC_ARCH }}${{ env.GCC_SUFFIX }}
+          tags: index.unikraft.io/kraftkit/gcc:${{ matrix.GCC_VERSION }}-${{ matrix.GCC_ARCH }}${{ env.GCC_SUFFIX }}
           platforms: linux/amd64
           secrets: |
             GIT_AUTH_TOKEN=${{ secrets.GITHUB_TOKEN }}
```

nderjung commented on .github/workflows/base.yaml:
```diff
@@ -0,0 +1,103 @@
+        uses: docker/build-push-action@v3
+        with:
+          file: ./buildenvs/qemu.Dockerfile
+          tags: kraftkit.sh/qemu:${{ QEMU_VERSION }}
```
```diff
diff --git a/.github/workflows/base.yaml b/.github/workflows/base.yaml
index ea6b450..94c6c1b 100644
--- a/.github/workflows/base.yaml
+++ b/.github/workflows/base.yaml
@@ -56,7 +56,7 @@ jobs:
         uses: docker/build-push-action@v3
         with:
           file: ./buildenvs/qemu.Dockerfile
-          tags: kraftkit.sh/qemu:${{ QEMU_VERSION }}
+          tags: index.unikraft.io/kraftkit/qemu:${{ QEMU_VERSION }}
           platforms: linux/amd64
           secrets: |
             GIT_AUTH_TOKEN=${{ secrets.GITHUB_TOKEN }}
```

nderjung commented on .github/workflows/base.yaml:
```diff
@@ -0,0 +1,103 @@
+        uses: docker/build-push-action@v3
+        with:
+          file: ./buildenvs/myself.Dockerfile
+          tags: kraftkit.sh/myself:latest
```
```diff
diff --git a/.github/workflows/base.yaml b/.github/workflows/base.yaml
index ea6b450..94c6c1b 100644
--- a/.github/workflows/base.yaml
+++ b/.github/workflows/base.yaml
@@ -74,7 +74,7 @@ jobs:
         uses: docker/build-push-action@v3
         with:
           file: ./buildenvs/myself.Dockerfile
-          tags: kraftkit.sh/myself:latest
+          tags: index.unikraft.io/kraftkit/myself:latest
           platforms: linux/amd64
           secrets: |
             GIT_AUTH_TOKEN=${{ secrets.GITHUB_TOKEN }}
```

nderjung commented on .github/workflows/base.yaml:
```diff
@@ -0,0 +1,103 @@
+        uses: docker/build-push-action@v3
+        with:
+          push: true
+          tags: kraftkit.sh/base:latest
```
```diff
diff --git a/.github/workflows/base.yaml b/.github/workflows/base.yaml
index ea6b450..94c6c1b 100644
--- a/.github/workflows/base.yaml
+++ b/.github/workflows/base.yaml
@@ -100,4 +100,4 @@ jobs:
         uses: docker/build-push-action@v3
         with:
           push: true
-          tags: kraftkit.sh/base:latest
+          tags: index.unikraft.io/kraftkit/base:latest
```

craciunoiuc committed:
Makefile: Proxy all container build commands

As a separate Makefile for handling builds already exists, all builds
are proxied to it to remove duplicated code. Thus, `make container` is
now `make buildenv-myself`.

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

nderjung commented on buildenvs/qemu.Dockerfile:
```diff
@@ -1,100 +1,201 @@
-COPY --from=qemu-build /bin/qemu-system-arm /bin/
-COPY --from=qemu-build /bin/qemu-system-i386 /bin/
-COPY --from=qemu-build /bin/qemu-system-x86_64 /bin/
+COPY --from=qemu-build /bin/qemu-system-x86_64 /bin/qemu-g[a] \
+    /bin/qemu-system-i38[6] /bin/qemu-system-ar[m] /bin/qemu-system-aarch6[4] \
+    /bin/qemu-pr-helpe[r] /bin/qemu-im[g] /bin/qemu-i[o] /bin/qemu-nb[d] /bin/
```
I believe this change was done to reduce the number of copy steps, right? I think because all other Dockerfiles are splitting things into new lines with a `\` separator we should do the same and we should be explicit with what we're copying, plus it makes diffs in the future easier to blame:
```diff
- COPY --from=qemu-build /bin/qemu-system-x86_64 /bin/qemu-g[a] \
-     /bin/qemu-system-i38[6] /bin/qemu-system-ar[m] /bin/qemu-system-aarch6[4] \
-     /bin/qemu-pr-helpe[r] /bin/qemu-im[g] /bin/qemu-i[o] /bin/qemu-nb[d] /bin/
+ COPY --from=qemu-build /bin/qemu-ga
+                        /bin/qemu-img \
+                        /bin/qemu-io \
+                        /bin/qemu-nbd \
+                        /bin/qemu-pr-helper \
+                        /bin/qemu-system-aarch64 \
+                        /bin/qemu-system-arm \
+                        /bin/qemu-system-i386 \
+                        /bin/qemu-system-x86_64 \
+                        /bin/
```

nderjung commented on buildenvs/Makefile:
```diff
@@ -0,0 +1,87 @@
+BUILDENVSDIR := $(patsubst %/,%,$(dir $(_MKFILE_PATH)))
+WORKDIR      ?= $(BUILDENVSDIR)/..
+
+REG          ?= kraftkit.sh
```
Just noticed this variable is inconsistent with the top-level Makefile.
```diff
- REG          ?= kraftkit.sh
+ REGISTRY     ?= kraftkit.sh
```

nderjung commented on buildenvs/Makefile:
```diff
@@ -0,0 +1,87 @@
+qemu: QEMU_VERSION ?= 7.1.0
+qemu: MAKE_NPROC   ?= $(shell nproc)
+qemu: ENVIRONMENT  ?= qemu
+qemu: IMAGE        ?= $(REG)/$(ENVIRONMENT):$(QEMU_VERSION)
```
```diff
- qemu: IMAGE        ?= $(REG)/$(ENVIRONMENT):$(QEMU_VERSION)
+ qemu: IMAGE        ?= $(REGISTRY)/$(ENVIRONMENT):$(QEMU_VERSION)
```

nderjung commented on buildenvs/Makefile:
```diff
@@ -0,0 +1,87 @@
+gcc: BINUTILS_VERSION ?= 2.39
+gcc: GLIB_VERSION     ?= 2.31
+gcc: ENVIRONMENT      ?= gcc
+gcc: IMAGE            ?= $(REG)/$(ENVIRONMENT):$(GCC_VERSION)-$(UK_ARCH)
```
```diff
- gcc: IMAGE            ?= $(REG)/$(ENVIRONMENT):$(GCC_VERSION)-$(UK_ARCH)
+ gcc: IMAGE            ?= $(REGISTRY)/$(ENVIRONMENT):$(GCC_VERSION)-$(UK_ARCH)
```

nderjung commented on buildenvs/Makefile:
```diff
@@ -0,0 +1,87 @@
+.PHONY: myself
+myself: GO_VERSION  ?= 1.20
+myself: ENVIRONMENT ?= myself
+myself: IMAGE       ?= $(REG)/$(ENVIRONMENT):$(IMAGE_TAG)
```
```diff
- myself: IMAGE       ?= $(REG)/$(ENVIRONMENT):$(IMAGE_TAG)
+ myself: IMAGE       ?= $(REGISTRY)/$(ENVIRONMENT):$(IMAGE_TAG)
```

nderjung commented on buildenvs/Makefile:
```diff
@@ -0,0 +1,87 @@
+base: GCC_PREFIX   ?= x86_64-linux-gnu
+base: QEMU_VERSION ?= 7.1.0
+base: ENVIRONMENT  ?= base
+base: IMAGE        ?= $(REG)/$(ENVIRONMENT):$(IMAGE_TAG)
+base:
+	$(DOCKER) build \
+		--build-arg REG=$(REG) \
```
```diff
- base: IMAGE        ?= $(REG)/$(ENVIRONMENT):$(IMAGE_TAG)
+ base: IMAGE        ?= $(REGISTRY)/$(ENVIRONMENT):$(IMAGE_TAG)
  base:
  	$(DOCKER) build \
- 		--build-arg REG=$(REG) \
+ 		--build-arg REG=$(REGISTRY) \
```

nderjung commented on buildenvs/base.Dockerfile:
```diff
@@ -0,0 +1,143 @@
+# Licensed under the BSD-3-Clause License (the "License").
+# You may not use this file except in compliance with the License.
+
+ARG REG=index.unikraft.io
+
+ARG UK_ARCH=x86_64
+
+ARG GCC_VERSION=12.2.0
+ARG GCC_SUFFIX=
+
+ARG QEMU_VERSION=7.1.0
+
+ARG KRAFTKIT_VERSION=latest
+
+ARG DEBIAN_VERSION=bullseye-20221114
```
Remove newline white space and switch registry to `kraftkit.sh`
```diff
- ARG REG=index.unikraft.io
+ ARG REGISTRY=kraftkit.sh
- 
  ARG UK_ARCH=x86_64
- 
  ARG GCC_VERSION=12.2.0
  ARG GCC_SUFFIX=
- 
  ARG QEMU_VERSION=7.1.0
- 
  ARG KRAFTKIT_VERSION=latest
- 
  ARG DEBIAN_VERSION=bullseye-20221114
```

nderjung commented on buildenvs/base.Dockerfile:
```diff
@@ -0,0 +1,143 @@
+
+ARG DEBIAN_VERSION=bullseye-20221114
+
+FROM ${REG}/gcc:${GCC_VERSION}-x86_64${GCC_SUFFIX} AS gcc-x86_64
+# FROM ${REG}/gcc:${GCC_VERSION}-arm${GCC_SUFFIX}    AS gcc-arm
+# FROM ${REG}/gcc:${GCC_VERSION}-arm64${GCC_SUFFIX}  AS gcc-arm64
+
+FROM ${REG}/qemu:${QEMU_VERSION}                   AS qemu
+FROM ${REG}/myself:${KRAFTKIT_VERSION}             AS kraft
```
```diff
- FROM ${REG}/gcc:${GCC_VERSION}-x86_64${GCC_SUFFIX} AS gcc-x86_64
+ FROM ${REGISTRY}/gcc:${GCC_VERSION}-x86_64${GCC_SUFFIX}  AS gcc-x86_64
- # FROM ${REG}/gcc:${GCC_VERSION}-arm${GCC_SUFFIX}    AS gcc-arm
+ # FROM ${REGISTRY}/gcc:${GCC_VERSION}-arm${GCC_SUFFIX}     AS gcc-arm
- # FROM ${REG}/gcc:${GCC_VERSION}-arm64${GCC_SUFFIX}  AS gcc-arm64
+ # FROM ${REGISTRY}/gcc:${GCC_VERSION}-arm64${GCC_SUFFIX}   AS gcc-arm64
- 
- FROM ${REG}/qemu:${QEMU_VERSION}                   AS qemu
+ FROM ${REGISTRY}/qemu:${QEMU_VERSION}                    AS qemu
- FROM ${REG}/myself:${KRAFTKIT_VERSION}             AS kraft
+ FROM ${REGISTRY}/myself:${KRAFTKIT_VERSION}              AS kraftkit
```

nderjung commented on .github/workflows/base.yaml:
```diff
@@ -0,0 +1,103 @@
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+
+      - name: Login to DockerHub
```
`Login to OCI registry`

nderjung commented on .github/workflows/gobuild.yaml:
```diff
@@ -0,0 +1,34 @@
+        with:
+          go-version: "1.20.x"
+
+      - name: Build Kraftkit
```
`Build KraftKit`

nderjung commented on buildenvs/base.Dockerfile:
```diff
@@ -0,0 +1,143 @@
+    apt-get clean; \
+    rm -Rf /var/cache/apt/*; \
+    rm -Rf /var/lib/apt/lists/*; \
+    mkdir -p /usr/src/unikraft/unikraft; \
+    mkdir -p /usr/src/unikraft/libs; \
+    mkdir -p /usr/src/unikraft/apps/app; \
```
Actually we can remove these since these are an artefact from `pykraft`:
```diff
-     mkdir -p /usr/src/unikraft/unikraft; \
-     mkdir -p /usr/src/unikraft/libs; \
-     mkdir -p /usr/src/unikraft/apps/app; \
```

nderjung commented on buildenvs/base.Dockerfile:
```diff
@@ -0,0 +1,143 @@
+FROM ${REG}/qemu:${QEMU_VERSION}                   AS qemu
+FROM ${REG}/myself:${KRAFTKIT_VERSION}             AS kraft
+
+# Copy the kraft tool
+COPY . /go/src/kraftkit.sh
+
+# Build the kraft tool - also installs it to the environment
+WORKDIR /go/src/kraftkit.sh
+
+RUN set -xe; \
+  make kraft;
```
Since we're `COPY`ing `--from=kraft` we don't need to run these steps:
```diff
- # Copy the kraft tool
- COPY . /go/src/kraftkit.sh
- 
- # Build the kraft tool - also installs it to the environment
- WORKDIR /go/src/kraftkit.sh
- 
- RUN set -xe; \
-   make kraft;
```

nderjung commented on buildenvs/base.Dockerfile:
```diff
@@ -0,0 +1,143 @@
+ARG GCC_PREFIX=x86_64-linux-gnu
+
+# Install kraftkit dependencies
+COPY --from=kraft /go/src/kraftkit.sh/dist/kraft /bin
```
```diff
- # Install kraftkit dependencies
- COPY --from=kraft /go/src/kraftkit.sh/dist/kraft /bin
+ # Install kraftkit distributables
+ COPY --from=kraftkit /go/src/kraftkit.sh/dist/kraft /usr/local/bin
```

nderjung commented on buildenvs/base.Dockerfile:
```diff
@@ -0,0 +1,143 @@
+# COPY --from=qemu /lib/arm-linux-gnueabihf/ /lib/arm-linux-gnueabihf
+# COPY --from=qemu /lib/aarch64-linux-gnu/ /lib/aarch64-linux-gnu
+
+# Unikraft workdir and environment
+WORKDIR /usr/src/unikraft/apps/app
```
This is an artefact from `pykraft`, we can assume the default debian working directory:
```diff
- # Unikraft workdir and environment
- WORKDIR /usr/src/unikraft/apps/app
```

nderjung commented on buildenvs/base.Dockerfile:
```diff
@@ -0,0 +1,143 @@
+# Unikraft workdir and environment
+WORKDIR /usr/src/unikraft/apps/app
+
+ENV LC_ALL=C.UTF-8
+ENV LANG=C.UTF-8
```
Since this is a container environment, let's also inject:
```diff
+ ENV KRAFTKIT_LOG_TYPE=basic
```

nderjung commented on buildenvs/myself.Dockerfile:
```diff
@@ -14,14 +14,28 @@ ARG GO_VERSION=${GO_VERSION}
+    make kraft; \
+    kraft -h;
+
+# Pull manifests
+RUN set -xe; \
+    kraft --log-level debug --log-type basic pkg source https://github.com/unikraft/app-*; \
+    kraft --log-level debug --log-type basic pkg update;
```
This should be removed, since it won't actually be used. The `myself` container is used to build and carry the binary artefacts from the KraftKit project, whereas the `base` container is used to actually build and run unikernels. This code should be migrated to the `base` container:
```diff
- # Pull manifests
- RUN set -xe; \
-     kraft --log-level debug --log-type basic pkg source https://github.com/unikraft/app-*; \
-     kraft --log-level debug --log-type basic pkg update;
```

nderjung commented on buildenvs/base.Dockerfile:
```diff
@@ -0,0 +1,143 @@
+    mkdir -p /usr/src/unikraft/unikraft; \
+    mkdir -p /usr/src/unikraft/libs; \
+    mkdir -p /usr/src/unikraft/apps/app; \
+    kraft -h;
```
The last step should actually include running kraft to gather manifests, and instead the `kraft -h` is used to only validate the `myself` container, as such:
```diff
-     kraft -h;
+     kraft --log-level debug --log-type basic pkg update;
```

nderjung commented on buildenvs/base.Dockerfile:
```diff
@@ -0,0 +1,143 @@
+RUN set -xe; \
+  make kraft;
+
+FROM debian:${DEBIAN_VERSION} AS kraftkit
```
```diff
- FROM debian:${DEBIAN_VERSION} AS kraftkit
+ FROM debian:${DEBIAN_VERSION}                      AS base
```

craciunoiuc committed:
.github/workflows: Add job to build 'myself-full'

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
.github/workflows: Fix go version to 1.20.2

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
.github/workflows: Make releases use kraftkit

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

nderjung reviewed (approved):
🚀 🚀 🚀 

Approved-by: Alexander Jung <alex@unikraft.io>
